airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -249,25 +279,139 @@ async def test_state_with_abnormally_large_values(
+        """
+        # If there is only one state message or less, the list is considered to be unique
+        if len(states) <= 1:
+            return states
+
+        stream_state_attr = "state.stream.stream_state"
+        record_count_attr = "state.sourceStats.recordCount"
+
+        # Define a function to get the stream state attribute from an AirbyteStateMessage object
+        get_state = operator.attrgetter(stream_state_attr)
+
+        # Define a function to get the record count attribute from an AirbyteStateMessage object
+        get_record_count = operator.attrgetter(record_count_attr)
+
+        current_idx = 0
+        result = []
+
+        # Iterate through the list of state messages
+        while current_idx < len(states) - 1:
+            next_idx = current_idx + 1
+            # Check if consecutive messages have the same stream state
+            while get_state(states[current_idx]) == get_state(states[next_idx]) and next_idx < len(states) - 1:
+                next_idx += 1
+
+            states[current_idx].state.sourceStats = AirbyteStateStats(recordCount=sum(map(get_record_count, states[current_idx:next_idx])))
+            # Append the first message with a unique stream state to the result list
+            result.append(states[current_idx])
+            # If the last message has a different stream state than the previous one, append it to the result list
+            if next_idx == len(states) - 1 and get_state(states[current_idx]) != get_state(states[next_idx]):
+                result.append(states[next_idx])
+            current_idx = next_idx
+
+        return list(self.get_expected_record_count_per_state(result))
+
+    def get_expected_record_count_per_state(self, states: List[AirbyteStateMessage]) -> Iterator[Tuple[AirbyteStateMessage, float]]:
+        """
+        Calculates the expected record count per state based on the total record count and distribution across states.
+        The expected record count is the number of records we expect to receive when applying a specific state checkpoint.
+
+        This function takes a list of state objects, calculates the total record count, and then
+        yields the expected record count for each state in reverse cumulative order. The expected
+        record count for each state is the total record count minus the sum of record counts of all
+        previous states in the list.
+
+        Args:
+            states (list): A non-empty list of state objects.
+
+        Yields:
+            tuple: A tuple containing a state object and its expected record count.
+
+        Raises:
+            Exception: If the 'states' list is empty.
+        """
+        if not states:
+            # Raise an exception if the 'states' list is empty
+            raise Exception("`states` expected to be a non empty list.")
+
+        # Function to extract the record count from a state object
+        get_record_count = operator.attrgetter("state.sourceStats.recordCount")
```

Comment:
nit: is this the simplest implementation? 

why can't we do something like

`sum([state.sourceStats.recordCount for state in states])`

In general, and in this case too, I think list comprehensions are easier to reason about than mapping functions

Reply:
I have prepared a function for this and wanted to reuse it, so we have a single point where we get the record count. After refactoring, get_record_count became a method and was reused multiple times in various methods.